adal/src/main/java/com/microsoft/aad/adal/ADALError.java

```diff
@@ -638,6 +634,38 @@ public String getLocalizedDescription(Context context) {
             return resources.getString(resources.getIdentifier(this.name(), "string",
                     context.getPackageName()));
         }
-        return mCommonADALError.getDescription();
+        return mDescription;
+    }
+    private static final Map<String, ADALError> exceptionMap = new HashMap<String,ADALError>();
+
+    static {
+        exceptionMap.put(CommonCoreException.ANDROIDKEYSTORE_FAILED, ADALError.ANDROIDKEYSTORE_FAILED);
+        exceptionMap.put(CommonCoreException.AUTHORITY_URL_NOT_VALID, ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_INSTANCE);
+        exceptionMap.put(CommonCoreException.AUTHORITY_VALIDATION_NOT_SUPPORTED, ADALError.DEVELOPER_AUTHORITY_CAN_NOT_BE_VALIDED);
+        exceptionMap.put(CommonCoreException.DECRYPTION_ERROR, ADALError.DECRYPTION_FAILED);
+        exceptionMap.put(CommonCoreException.DEVICE_NETWORK_NOT_AVAILABLE, ADALError.DEVICE_CONNECTION_IS_NOT_AVAILABLE);
+        exceptionMap.put(CommonCoreException.ENCRYPTION_ERROR, ADALError.ENCRYPTION_ERROR);
+        exceptionMap.put(CommonCoreException.INVALID_JWT, ADALError.JSON_PARSE_ERROR);
+        exceptionMap.put(CommonCoreException.IO_ERROR, ADALError.IO_EXCEPTION);
+        exceptionMap.put(CommonCoreException.JSON_PARSE_FAILURE, ADALError.JSON_PARSE_ERROR);
+        // TODO: :O
+        exceptionMap.put(CommonCoreException.MALFORMED_URL, ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_URL);
+        exceptionMap.put(CommonCoreException.MULTIPLE_MATCHING_TOKENS_DETECTED, ADALError.AUTH_FAILED_USER_MISMATCH);
+        exceptionMap.put(CommonCoreException.USER_MISMATCH, ADALError.AUTH_FAILED_USER_MISMATCH);
+        exceptionMap.put(CommonCoreException.NO_SUCH_ALGORITHM, ADALError.DEVICE_NO_SUCH_ALGORITHM);
+        exceptionMap.put(CommonCoreException.UNSUPPORTED_ENCODING, ADALError.ENCODING_IS_NOT_SUPPORTED);
+
+        // The following mapping need a lot more brushing up
+        exceptionMap.put(CommonCoreServiceException.ACCESS_DENIED, ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED);
+        exceptionMap.put(CommonCoreServiceException.INVALID_INSTANCE, ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED);
+        exceptionMap.put(CommonCoreServiceException.INVALID_REQUEST, ADALError.SERVER_INVALID_REQUEST);
+        exceptionMap.put(CommonCoreServiceException.REQUEST_TIMEOUT, ADALError.SERVER_ERROR);
+        exceptionMap.put(CommonCoreServiceException.SERVICE_NOT_AVAILABLE, ADALError.SERVER_ERROR);
+        exceptionMap.put(CommonCoreServiceException.UNKNOWN_ERROR, ADALError.SERVER_ERROR);
+    }
+
+    public static ADALError fromCommon(final CommonCoreBaseException exception) {
+        // What do we do if errorCode is null?
```

Comment:
ADAL did not use AuthenticationException with null errorCode, although it does not have a null check in the constructor.  I think the question should be in which circumstance we will also `null` erroCode in CommonCoreBaseException or we should not. 

There is another question. Which if the map does not contains the key-value pair for CommonCoreBaseException.getErrorCode() mapping to the ADALError? In the current implementation, we need to add the mapping when we introduce a new errorCode in common-core. And it is hard to find the mapping miss until the CommonCoreBaseException is thrown. As a result, we probably return the AuthenticationException with null ADAL ErrorCode. Should we also take into consideration of this case?